allennlp/training/trainer.py

```diff
@@ -507,11 +507,11 @@ def _should_stop_early(self, metric_history: List[float]) -> bool:
         uses patience and the validation metric to determine if training should stop early
         """
         if len(metric_history) > self._patience:
-            # Is the best score in the past N epochs worse than the best score overall?
+            # Is the best score in the past N epochs worse than or equal the best score overall?
             if self._validation_metric_decreases:
-                return min(metric_history[-self._patience:]) > min(metric_history)
+                return min(metric_history[-self._patience:]) >= min(metric_history[:-self._patience])
```

Comment:
Regarding https://github.com/allenai/allennlp/issues/1267, for ``patience == 0`` the proposed code crashes: it takes the min of the empty list ``metric_history[:-0]``.

Reply:
From a functional point of view what would patience=0 mean?
intuitively it would read as "one of the last 0 entries needs to be better than all historical results?"  This obviously doesn't make any sense. 
I don't think ppl should be setting patience to 0 ever. 
Do you agree? 
I will add a guard in the constructor to protect against this.

Reply:
Thanks! I'm going to wait for your PR to get merged and then I will pick this back up.

Reply:
Actually, I hadn't in mind to open a PR :D but, well, it wasn't  a lot of work so I did: https://github.com/allenai/allennlp/pull/1271